### PR TITLE
scatter plot and hexbin plot lose x-axis when colorbar is included.

### LINKED ISSUE #10611: xticks missing for scatter plots with colors
example: https://www.wakari.io/sharing/bundle/adamgreenhall/test-scatter

I think this happens specifically for pandas scatter plots with colorbars in ipython. The xticks are still working for:
- non-colorbar pandas scatter plots
- the same scatter plot using matplotlib
- standard python scripts using `plt.savefig`

related problem with `%matplotlib inline`?: https://github.com/ipython/ipython/issues/1443/

### LINKED ISSUE #10678: Hexbin plots does not display x label and xtick labels
I'm currently using python 2.7, pandas 0.16.2.dev, and ipython version 3.1.0

In brief, it seems that there are cases that the sharex option of pandas.DataFrame.plot(kind='hexbin') defaults to True instead of False. This causes the x label and xtick labels to be invisible.

pandas-sharex-bug

Preceding stack overflow and ipython github threads that led to here:

http://stackoverflow.com/questions/31220329/pandas-dataframe-hexbin-plot-has-no-xlabel-or-axis-values

https://github.com/ipython/ipython/issues/8653

### PR DESCRIPTION
**Pull request by javadnoorb**
closes https://github.com/pandas-dev/pandas/issues/10611 
closes https://github.com/pandas-dev/pandas/issues/10678. 

The x-axis for scatter plot and hexbin plot disappears when colorbar is included. This seems to be because colorbar axis is looped through in `_handle_shared_axes`: https://github.com/pandas-dev/pandas/blob/8a5830305d2d9a087fbe46bd968218104ffdfc49/pandas/plotting/_core.py#L426
after discussing with @TomAugspurger (issue https://github.com/pandas-dev/pandas/issues/10611), we decided to try adding the attribute `__is_pandas_colorbar` to colorbar axis object and skipping it during handling of shared axes. 

I've done some tests that seem to fix the issue. But we may need more tests:
```
%matplotlib inline
import matplotlib.pylab as pl
from mypandas import pandas as pd
import numpy as np
random_array = np.random.random((1000,3))

df = pd.DataFrame(random_array,columns=['A label','B label','C label'])
df.plot.scatter('A label','B label',c='C label', PATCH_MODE_FLAG = False);pl.title('pandas current version');
df.plot.scatter('A label','B label',c='C label', PATCH_MODE_FLAG = True);pl.title('patch fixing x-axis');

df.plot.hexbin('A label','B label', gridsize=25, PATCH_MODE_FLAG = False);pl.title('pandas current version');
df.plot.hexbin('A label','B label', gridsize=25, PATCH_MODE_FLAG = True);pl.title('patch fixing x-axis');
```

image
image
image
image

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thanks, let us know when you're ready for us to take a look.

How does this work with `subplots=True`? Do we not have to worry about that since it's a scatter plot?

How does this work with user-specified axes via `ax=`? Especially with multiple subplots. It may already be broken, so no worries if it doesn't work well.

**Comment by javadnoorb** [comment]
Yep, it's already broke. Please feel free to take a look:

```
def make_multiplots(numrows=2,numcols=2,sharey=True,sharex=True,PATCH_MODE_FLAG = True, is_scatter = True):
    fig, axes = pl.subplots(numrows,numcols,sharey=sharey,sharex=sharex )
    fig.set
    for i in range(axes.shape[0]):
        for j in range(axes.shape[1]):
            random_array = np.random.random((1000,3))
            df = pd.DataFrame(random_array,columns=['A label','B label','C label'])
            if is_scatter:
                df.plot.scatter('A label','B label',c='C label', PATCH_MODE_FLAG = PATCH_MODE_FLAG,ax=axes[i,j]);
            else:
                df.plot.hexbin('A label','B label',gridsize=20, PATCH_MODE_FLAG = PATCH_MODE_FLAG,ax=axes[i,j]);
            
            
make_multiplots(numrows=2,numcols=2,sharey=True, sharex=True, PATCH_MODE_FLAG = False, is_scatter=True);pl.suptitle('pandas current version');            
make_multiplots(numrows=2,numcols=2,sharey=True, sharex=True, PATCH_MODE_FLAG = True, is_scatter=True);pl.suptitle('patch fixing x-axis');            
make_multiplots(numrows=2,numcols=2,sharey=True, sharex=True, PATCH_MODE_FLAG = False, is_scatter=False);pl.suptitle('pandas current version');            
make_multiplots(numrows=2,numcols=2,sharey=True, sharex=True, PATCH_MODE_FLAG = True, is_scatter=False);pl.suptitle('patch fixing x-axis');  
```          
image
image
image
image

File: pandas/plotting/_core.py [hunk 1]
```diff
@@ -870,6 +870,8 @@ def _make_plot(self):
         scatter = ax.scatter(data[x].values, data[y].values, c=c_values,
                              label=label, cmap=cmap, **self.kwds)
         if cb:
+            if PATCH_MODE:
```
Comment (TomAugspurger) [inline pandas/plotting/_core.py]: I assume this ``PATCH_MODE` is just for debugging?

File: pandas/plotting/_tools.py [hunk 2]
```diff
@@ -298,20 +298,20 @@ def _remove_labels_from_axis(axis):
 
 def _handle_shared_axes(axarr, nplots, naxes, nrows, ncols, sharex, sharey):
     if nplots > 1:
-
+
```
Comment (TomAugspurger) [inline pandas/plotting/_tools.py]: This whitespace will likely fail the linter

File: pandas/plotting/_tools.py [hunk 3]
```diff
@@ -298,20 +298,20 @@ def _remove_labels_from_axis(axis):
...
+        
         if nrows > 1:
             try:
                 # first find out the ax layout,
                 # so that we can correctly handle 'gaps"
                 layout = np.zeros((nrows + 1, ncols + 1), dtype=np.bool)
                 for ax in axarr:
                     layout[ax.rowNum, ax.colNum] = ax.get_visible()
-
+
```
Comment (TomAugspurger) [inline pandas/plotting/_tools.py]: Likewise

**Review by TomAugspurger — COMMENTED** [review]
I would say make a new issue for subplots, and ignore it for now.

As for testing, how about we make two plots, one with `c` and one without, and assert that the tick labels are the same?


```
fig, ax = plt.supblots()
df.plot.scatter(x='x', y='y', ax=ax)

fig, ax2 = plt.subplots()
df.plot.scatter(x='x', y='y', c='c', ax=ax2)

assert ax.get_xticklabels() == ax2.get_xticklabels()
```

or something like that.

**Comment by javadnoorb** [comment]
As of subplots=True, I've only used it with histograms and line plots. I'm not familiar with their usage with scatter plots/hexbin plots. So I'm not sure how to test them. Simply using it with single plot doesn't interfere with the outcome, but I'm assuming that's not what you were referring to. I think we'll need your expertise here.

`df.plot.scatter('A label','B label',c='C label', PATCH_MODE_FLAG = True, subplots=True);`
image

File: pandas/plotting/_core.py (hunk 1 above)
Comment (javadnoorb) [reply pandas/plotting/_core.py]: yes, it should be removed. Didn't know of a better way to do this.

**Review by javadnoorb — COMMENTED** [review]
removed the whitespaces

**Comment by TomAugspurger** [comment]
Don't worry about subplots. It would be something like `df.plot.scatter(x=0, y=[1, 2], subplots=True)`, which currently isn't allowed.

**Comment by javadnoorb** [comment]
That test with `assert` doesn't deal with the issue. It seems like if you make a subplot and pass the axis to pandas, the x-axis comes out fine. I didn't use environments, and don't know how to switch from my development pandas to the native version without an uninstall. But with the `PATCH_MODE_FLAG=False` you can see it:

```
fig, ax1 = pl.subplots()
df.plot.scatter(x='A label', y='B label', c='C label',ax=ax1, PATCH_MODE_FLAG=False)
```
image


`df.plot.scatter(x='A label', y='B label', c='C label', PATCH_MODE_FLAG=False)`
image

**Comment by javadnoorb** [comment]
This works:

```
PATCH_MODE_FLAG = False
ax1 = df.plot.scatter(x='A label', y='B label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
ax2 = df.plot.scatter(x='A label', y='B label', c='C label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
assert (ax1.get_xticks()==ax2.get_xticks()).all(), "axis labels are not the same"
```

but regardless of `PATCH_MODE_FLAG` value it doesn't throw an error. 

I think the issue is not with the xticks attribute but rather showing them.

**Comment by javadnoorb** [comment]
Trying a few of these. Quite bizarre:

```
PATCH_MODE_FLAG = False
ax1 = df.plot.scatter(x='A label', y='B label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
ax2 = df.plot.scatter(x='A label', y='B label', c='C label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
assert (ax1.get_xticks()==ax2.get_xticks()).all(), "Axis xticks are not the same"
assert ax1.xaxis.get_visible()==ax2.xaxis.get_visible(), "x-axis visbilities are not the same"
print(ax1.get_xticklabels(),ax2.get_xticklabels())
print(ax1.get_xticks(),ax2.get_xticks())
print(ax1.xaxis.get_visible(),ax2.xaxis.get_visible())
```

image

**Comment by TomAugspurger** [comment]
@javadnoorb how's this going? For this PR I would ignore user-provided axes, and only worry about the common case of `df.plot(x='x', y='y', c='c')`.

**Comment by javadnoorb** [comment]
@TomAugspurger I haven't worked on this for a few days. Let me see if I can come up with a reasonable test. `get_xticklabels` returns an object, so I think we can't use it for comparison. I think the reason the x-labels disappear should be because of these two lines:

https://github.com/pandas-dev/pandas/blob/8a5830305d2d9a087fbe46bd968218104ffdfc49/pandas/plotting/_tools.py#L282
https://github.com/pandas-dev/pandas/blob/8a5830305d2d9a087fbe46bd968218104ffdfc49/pandas/plotting/_tools.py#L296

my guess is that something like:

`assert ax1.get_majorticklabels().get_visible()==ax2.get_majorticklabels().get_visible()` 

and 

`assert ax1.get_label().get_visible()==ax2.get_label().get_visible()` 

or something like that, should work for the test. 

Let me give it a shot.

**Comment by TomAugspurger** [comment]
Sounds good. Let me know if that doesn't work and I can take a look.

On Wed, Mar 28, 2018 at 11:48 AM, Javad Noorbakhsh <notifications@github.com

**Comment by javadnoorb** [comment]
Ok, I wrote a test that seems to work:
```
def test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG=True):
    random_array = np.random.random((1000,3))
    df = pd.DataFrame(random_array,columns=['A label','B label','C label'])

    ax1 = df.plot.scatter(x='A label', y='B label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
    ax2 = df.plot.scatter(x='A label', y='B label', c='C label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)

    assert all([vis[0].get_visible()==vis[1].get_visible() for vis in zip(ax1.xaxis.get_minorticklabels(),ax2.xaxis.get_minorticklabels())]), 'minor x-axis tick labels visibility changes when colorbar included'
    assert all([vis[0].get_visible()==vis[1].get_visible() for vis in zip(ax1.xaxis.get_majorticklabels(),ax2.xaxis.get_majorticklabels())]), 'major x-axis tick labels visibility changes when colorbar included'
    assert ax1.xaxis.get_label().get_visible()==ax2.xaxis.get_label().get_visible(),'x-axis label visibility changes when colorbar included'
```
```    
test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG=True)    
```
image
image
```
test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG=False)
```

```---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-47-0cc25b4e00f4> in <module>()
     10     assert ax1.xaxis.get_label().get_visible()==ax2.xaxis.get_label().get_visible(),'x-axis label visibility changes when colorbar included'
     11 
---> 12 test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG=False)

<ipython-input-47-0cc25b4e00f4> in test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG)
      7 
      8     assert all([vis[0].get_visible()==vis[1].get_visible() for vis in zip(ax1.xaxis.get_minorticklabels(),ax2.xaxis.get_minorticklabels())]), 'minor x-axis tick labels visibility changes when colorbar included'
----> 9     assert all([vis[0].get_visible()==vis[1].get_visible() for vis in zip(ax1.xaxis.get_majorticklabels(),ax2.xaxis.get_majorticklabels())]), 'major x-axis tick labels visibility changes when colorbar included'
     10     assert ax1.xaxis.get_label().get_visible()==ax2.xaxis.get_label().get_visible(),'x-axis label visibility changes when colorbar included'
     11 

AssertionError: major x-axis tick labels visibility changes when colorbar included
```

image
image


If that seems fine I'll include the changes and remove the PATCH_MODE flag. 

I'm not familiar with writing tests. Is this where it goes?
https://github.com/pandas-dev/pandas/blob/master/pandas/tests/plotting/test_misc.py

**Comment by TomAugspurger** [comment]
That test seems sensible.

Tests can go in `pandas/pandas/tests/plotting/test_frame.py` under TestFrame. There are some other ones there for `scatter`.

**Comment by javadnoorb** [comment]
Ok, great. I'll check if it passes the regular checks before including the test.

**Comment by javadnoorb** [comment]
Also need a test for hexbin. This one seems to work:

```
def test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG=True):
    random_array = np.random.random((1000,3))
    df = pd.DataFrame(random_array,columns=['A label','B label','C label'])

    ax = df.plot.hexbin('A label','B label', gridsize=12, PATCH_MODE_FLAG = PATCH_MODE_FLAG);    
    assert all([vis.get_visible() for vis in ax.xaxis.get_minorticklabels()]), 'minor x-axis tick labels are not visible'
    assert all([vis.get_visible() for vis in ax.xaxis.get_majorticklabels()]), 'major x-axis tick labels are not visible'
    assert ax.xaxis.get_label().get_visible(),'x-axis label is not visible'
```

```
test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG=True)   
```
image

```
test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG=False)   
```
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-66-cd06ece40a4d> in <module>()
      9 
     10 
---> 11 test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG=False)

<ipython-input-66-cd06ece40a4d> in test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG)
      4 
      5     ax = df.plot.hexbin('A label','B label', gridsize=12, PATCH_MODE_FLAG = PATCH_MODE_FLAG);
----> 6     assert all([vis.get_visible() for vis in ax.xaxis.get_minorticklabels()]), 'minor x-axis tick labels are not visible'
      7     assert all([vis.get_visible() for vis in ax.xaxis.get_majorticklabels()]), 'major x-axis tick labels are not visible'
      8     assert ax.xaxis.get_label().get_visible(),'x-axis label is not visible'

AssertionError: minor x-axis tick labels are not visible
```
image

**Comment by javadnoorb** [comment]
These changes seem pretty benign. Is there any specific reason for Travis-CI to throw an error?

File: pandas/plotting/_tools.py [hunk 4]
```diff
@@ -311,7 +311,7 @@ def _handle_shared_axes(axarr, nplots, naxes, nrows, ncols, sharex, sharey):
                     # only the last row of subplots should get x labels -> all
                     # other off layout handles the case that the subplot is
                     # the last in the column, because below is no subplot/gap.
-                    if not layout[ax.rowNum + 1, ax.colNum]:
+                    if not layout[ax.rowNum + 1, ax.colNum] or getattr(ax, '_pandas_colorbar_axes', False):
```
Comment (TomAugspurger) [inline pandas/plotting/_tools.py]: This line is too long: https://travis-ci.org/pandas-dev/pandas/jobs/359516939#L3006

Comment (javadnoorb) [reply pandas/plotting/_tools.py]: Thanks. I'll fix it. Sorry I'm pretty new to the open-source world.

File: pandas/plotting/_core.py [hunk 5]
```diff
@@ -870,6 +870,7 @@ def _make_plot(self):
         scatter = ax.scatter(data[x].values, data[y].values, c=c_values,
                              label=label, cmap=cmap, **self.kwds)
         if cb:
+            ax._pandas_colorbar_axes = True
```
Comment (TomAugspurger) [inline pandas/plotting/_core.py]: Add a comment here explaining why you're adding this attribute.

File: pandas/plotting/_core.py [hunk 6]
```diff
@@ -915,6 +917,7 @@ def _make_plot(self):
         ax.hexbin(data[x].values, data[y].values, C=c_values, cmap=cmap,
                   **self.kwds)
         if cb:
+            ax._pandas_colorbar_axes = True
```
Comment (TomAugspurger) [inline pandas/plotting/_core.py]: Comment.

File: pandas/plotting/_tools.py [hunk 7]
```diff
@@ -311,7 +311,8 @@ def _handle_shared_axes(axarr, nplots, naxes, nrows, ncols, sharex, sharey):
                     # only the last row of subplots should get x labels -> all
                     # other off layout handles the case that the subplot is
                     # the last in the column, because below is no subplot/gap.
-                    if not layout[ax.rowNum + 1, ax.colNum]:
+                    if not layout[ax.rowNum + 1, ax.colNum] or \
```
Comment (TomAugspurger) [inline pandas/plotting/_tools.py]: We use parenthesis to break long lines, not `\`.

File: pandas/tests/plotting/test_frame.py [hunk 8]
```diff
@@ -1030,6 +1030,43 @@ def test_plot_scatter(self):
         axes = df.plot(x='x', y='y', kind='scatter', subplots=True)
         self._check_axes_shape(axes, axes_num=1, layout=(1, 1))
 
+    @pytest.mark.slow
+    def test_if_scatterplot_colorbar_affects_xaxis_visibility(self):
+        random_array = np.random.random((1000,3))
```
Comment (TomAugspurger) [inline pandas/tests/plotting/test_frame.py]: PEP8.

You can run `flake8 pandas/tests/plotting/test_frame.py` locally to check this.

File: pandas/tests/plotting/test_frame.py [hunk 9]
```diff
@@ -1030,6 +1030,43 @@ def test_plot_scatter(self):
...
+        df = pd.DataFrame(random_array,columns=['A label','B label','C label'])
+
+        ax1 = df.plot.scatter(x='A label', y='B label')
+        ax2 = df.plot.scatter(x='A label', y='B label', c='C label')
+
+        assert all([vis[0].get_visible() == vis[1].get_visible() for vis in
```
Comment (TomAugspurger) [inline pandas/tests/plotting/test_frame.py]: Just do the `assert`, no error message. pytest will ensure that if the test fails, we still get a meaningful error message.

Comment (TomAugspurger) [reply pandas/tests/plotting/test_frame.py]: Further, I think you can just do

```python
vis1 = [vis.get_visible() for vis in ax1.xaxis.get_minorticklabels()]
vis2 = [vis.get_visible() for vis in ax2.xaxis.get_minorticklabels()]

assert vis1 == vis2
```

As written, if one happens to be shorter than the other, we wouldn't know because the `zip` would just stop.

File: pandas/tests/plotting/test_frame.py (hunk 8 above)
Comment (TomAugspurger) [inline pandas/tests/plotting/test_frame.py]: Include a reference to the github issue as a comment.

**Review by TomAugspurger — COMMENTED** [review]
Good start. Could you include a release note under the bugfix section (I think we have a vis subsection). 0.23.0.txt.

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/blob/master/doc/source/whatsnew/v0.23.0.txt

**Comment by javadnoorb** [comment]
I've been trying the tests with the inline and pdf backend. Tests pass depending on the backend. So I'm guessing that my PR doesn't really address the issue.

**Comment by TomAugspurger** [comment]
> Tests pass depending on the backend. So I'm guessing that my PR doesn't really address the issue.

Can you rephrase that? This behavior only manifests with specific backends. Ideally it would always fail on a given backend (e.g. notebook) without your changes, and always pass with it.

If the backend is too much of an issue, we can adapt the test. The change is clearly an improvement.

**Comment by javadnoorb** [comment]
@TomAugspurger. Seems like the issue is because in the inline backend floats corresponding to location of axes have too many arbitrary trailing digits. For example two axes that have their bottom left corners at 0.3 may appear as 0.3 and 0.30000001. I rounded them in `_core.py`.

This new commit works for single plots with pdf and inline backends. It still fails for multiplots. We'll need to dig deeper for that. But it seems like we don't need the previous hack anymore.

File: pandas/plotting/_core.py [hunk 10]
```diff
@@ -788,6 +788,11 @@ def _get_axes_layout(self):
         for ax in axes:
             # check axes coordinates to estimate layout
             points = ax.get_position().get_points()
+            # in IPython inline backend, floats returned by
+            #  `get_points()` have too many ad hoc trailing
+            # digits. Unless rounded these values won't
+            # match in the following set operations
+            points = np.round(points, 3)
```
Comment (TomAugspurger) [inline pandas/plotting/_core.py]: I worry about this causing debugging nightmares further down the road... I'd prefer to solve this local to the problem if possible.

If not, you could perhaps include a keyword to `_get_axes_layout`? I'm not sure how difficult that would be to plumb through.

Comment (javadnoorb) [reply pandas/plotting/_core.py]: I agree. That's pretty tough to track. 

I made a new commit. The changes are local now. I'm forcing colorbars to take on their relevant axes positions from the parent axes. The merge passes the tests and resolves the xlabel issue (#10611 and #10678) but not with the subplots (#20455).

**Comment by javadnoorb** [comment]
I don't know what happened with Travis-CI, but looks like there was a connection problem in the middle of the run. I tried this on my own fork and Travis only threw two errors instead (pasted below). It looks like `read_html` is trying to read a dead link for testing:

http://ndb.nal.usda.gov/ndb/foods/show/1732?fg=&man=&lfacet=&format=&count=&max=25&offset=&sort=&qlookup=spam

Any ideas what happened?


```
=================================== FAILURES ===================================
_______________________ TestReadHtml.test_spam_url[bs4] ________________________
[gw1] linux2 -- Python 2.7.14 /home/travis/miniconda3/envs/pandas/bin/python
self = <pandas.tests.io.test_html.TestReadHtml object at 0x7f2989916310>
    @network
    def test_spam_url(self):
        url = ('http://ndb.nal.usda.gov/ndb/foods/show/1732?fg=&man=&'
               'lfacet=&format=&count=&max=25&offset=&sort=&qlookup=spam')
>       df1 = self.read_html(url, '.*Water.*')
pandas/tests/io/test_html.py:106: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/io/html.py:983: in read_html
    displayed_only=displayed_only)
pandas/io/html.py:811: in _parse
    raise_with_traceback(retained)
pandas/io/html.py:793: in _parse
    tables = p.parse_tables()
pandas/io/html.py:213: in parse_tables
    tables = self._parse_tables(self._build_doc(), self.match, self.attrs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <pandas.io.html._BeautifulSoupHtml5LibFrameParser object at 0x7f2989598310>
doc = <!DOCTYPE html>\n<!--[if lt IE 7 ]> <html lang="en" class="no-js ie6"> <![endi...t\t</div>\n\t\t</small>\n\t\t</footer>\n\t\t\n\t\t\n    \n   \n\n</body></html>
match = <_sre.SRE_Pattern object at 0x7f29899c2b20>, attrs = None
    def _parse_tables(self, doc, match, attrs):
        element_name = self._strainer.name
        tables = doc.find_all(element_name, attrs=attrs)
    
        if not tables:
>           raise ValueError('No tables found')
E           ValueError: No tables found
pandas/io/html.py:467: ValueError
_______________________ TestReadHtml.test_spam_url[lxml] _______________________
[gw1] linux2 -- Python 2.7.14 /home/travis/miniconda3/envs/pandas/bin/python
self = <pandas.tests.io.test_html.TestReadHtml object at 0x7f298ca9fe10>
    @network
    def test_spam_url(self):
        url = ('http://ndb.nal.usda.gov/ndb/foods/show/1732?fg=&man=&'
               'lfacet=&format=&count=&max=25&offset=&sort=&qlookup=spam')
>       df1 = self.read_html(url, '.*Water.*')
pandas/tests/io/test_html.py:106: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/io/html.py:983: in read_html
    displayed_only=displayed_only)
pandas/io/html.py:811: in _parse
    raise_with_traceback(retained)
pandas/io/html.py:793: in _parse
    tables = p.parse_tables()
pandas/io/html.py:213: in parse_tables
    tables = self._parse_tables(self._build_doc(), self.match, self.attrs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <pandas.io.html._LxmlFrameParser object at 0x7f298ca9f610>
doc = <Element html at 0x7f298c980730>
match = <_sre.SRE_Pattern object at 0x7f29899c2b20>, kwargs = None
    def _parse_tables(self, doc, match, kwargs):
        pattern = match.pattern
    
        # 1. check all descendants for the given pattern and only search tables
        # 2. go up the tree until we find a table
        query = '//table//*[re:test(text(), {patt!r})]/ancestor::table'
        xpath_expr = u(query).format(patt=pattern)
    
        # if any table attributes were given build an xpath expression to
        # search for them
        if kwargs:
            xpath_expr += _build_xpath_expr(kwargs)
    
        tables = doc.xpath(xpath_expr, namespaces=_re_namespace)
    
        tables = self._handle_hidden_tables(tables, "attrib")
        if self.displayed_only:
            for table in tables:
                # lxml utilizes XPATH 1.0 which does not have regex
                # support. As a result, we find all elements with a style
                # attribute and iterate them to check for display:none
                for elem in table.xpath('.//*[@style]'):
                    if "display:none" in elem.attrib.get(
                            "style", "").replace(" ", ""):
                        elem.getparent().remove(elem)
    
        if not tables:
            raise ValueError("No tables found matching regex {patt!r}"
>                            .format(patt=pattern))
E           ValueError: No tables found matching regex '.*Water.*'
pandas/io/html.py:586: ValueError
```

**Comment by javadnoorb** [comment]
I just remembered about this. The last commit seemed to have passed all the checks. Is there any interest to follow up on it?

File: doc/source/whatsnew/v0.23.0.txt [hunk 11]
```diff
@@ -1176,7 +1176,7 @@ Plotting
 - Bug in formatting tick labels with ``datetime.time()`` and fractional seconds (:issue:`18478`).
 - :meth:`Series.plot.kde` has exposed the args ``ind`` and ``bw_method`` in the docstring (:issue:`18461`). The argument ``ind`` may now also be an integer (number of sample points).
 - :func:`DataFrame.plot` now supports multiple columns to the ``y`` argument (:issue:`19699`)
-
+- Bug in :func:'DataFrame.plot.scatter' and :func:'DataFrame.plot.hexbin' caused x-axis label and ticklabels to disappear when colorbar was on in IPython inline backend (:issue:`10611` and :issue:`10678`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: This should be moved to 0.24.0

File: pandas/plotting/_core.py [hunk 12]
```diff
@@ -882,8 +882,13 @@ def _make_plot(self):
             kws = dict(ax=ax)
             if self.mpl_ge_1_3_1():
                 kws['label'] = c if c_is_column else ''
-            self.fig.colorbar(img, **kws)
-
+            cbar = self.fig.colorbar(img, **kws)
```
Comment (TomAugspurger) [inline pandas/plotting/_core.py]: Maybe refactor this and https://github.com/pandas-dev/pandas/pull/20446/files#diff-2b118bda866f3d626ceb6b529c62cd1aR932 to a method. They're identical, right? I guess aside from `**kws` vs. `ax=ax`?

**Review by TomAugspurger — COMMENTED** [review]
Sorry for dropping this. Can you merge in master and move the whatsnew to 0.24.0?

Looks good otherwise, so ping on green.

File: pandas/plotting/_core.py [hunk 13]
```diff
@@ -833,6 +833,23 @@ def _post_plot_logic(self, ax, data):
         ax.set_ylabel(pprint_thing(y))
         ax.set_xlabel(pprint_thing(x))
 
+    def _plot_colorbar(self, kws):
```
Comment (jreback) [inline pandas/plotting/_core.py]: can you make ax a named arg here and so you would pass as ``**kwds`` when you call.

File: pandas/plotting/_core.py [hunk 14]
```diff
@@ -833,6 +833,23 @@ def _post_plot_logic(self, ax, data):
...
+        # In IPython inline backend the colorbar axis height
+        # tends to not exactly match the parent axis height.
+        # The difference is due to small fractional differences
+        # in floating points with similar representation.
+        # To deal with this, this method forces the colorbar
+        # height to take the height of the parent axes.
+        ax = kws['ax']
+        img = ax.collections[0]
```
Comment (jreback) [inline pandas/plotting/_core.py]: is there an issue we can reference about this impl. IOW next person to look at wouldn't know where these set_position come from. so a) ref this issue in a comment and b) an Ipython refernce (if you can)

**Review by jreback — CHANGES_REQUESTED** [review]
comments, otherwise lgtm. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @javadnoorb very nice!

**Comment by javadnoorb** [comment]
thanks @jreback!
